pkg/storage/storagewrappers/metrics.go

```diff
@@ -0,0 +1,73 @@
+package storagewrappers
+
+import (
+	"context"
+	"sync/atomic"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/pkg/storage"
+)
+
+var _ storage.RelationshipTupleReader = (*MetricsOpenFGAStorage)(nil)
+
+type MetricsOpenFGAStorage struct {
+	storage.RelationshipTupleReader
+	counter atomic.Uint32
+}
+
+func NewMetricsOpenFGAStorage(wrapped storage.RelationshipTupleReader) *MetricsOpenFGAStorage {
+	return &MetricsOpenFGAStorage{
+		RelationshipTupleReader: wrapped,
+		counter:                 atomic.Uint32{},
+	}
+}
+
+type Metrics struct {
+	DatastoreQueryCount int
+}
+
+func (m *MetricsOpenFGAStorage) GetMetrics() Metrics {
+	return Metrics{
+		DatastoreQueryCount: int(m.counter.Load()),
+	}
+}
+
+func (m *MetricsOpenFGAStorage) increase() {
+	m.counter.Add(1)
+}
+
+// Read see [storage.RelationshipTupleReader.ReadUserTuple].
+func (m *MetricsOpenFGAStorage) Read(ctx context.Context, store string, tupleKey *openfgav1.TupleKey, options storage.ReadOptions) (storage.TupleIterator, error) {
+	defer m.increase()
```

Comment:
Why use `defer`? It's just an addition to the call stack with no point 🤔 More operations overall to process.